server/src/main/java/io/crate/expression/tablefunctions/MatchesFunction.java

```diff
@@ -0,0 +1,208 @@
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING_ARRAY.getTypeSignature()
+            ),
+            MatchesFunction::new
+        );
+    }
+
+    @Nullable
+    private Pattern pattern;
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private MatchesFunction(Signature signature, Signature boundSignature) {
+        this(signature, boundSignature, null);
+    }
+
+    private MatchesFunction(Signature signature, Signature boundSignature, @Nullable Pattern pattern) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        this.pattern = pattern;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public RowType returnType() {
+        return ROW_TYPE;
+    }
+
+    @Override
+    public boolean hasLazyResultSet() {
+        return false;
+    }
+
+    @VisibleForTesting
+    Pattern pattern() {
+        return pattern;
+    }
+
+    @Override
+    public Scalar<Iterable<Row>, List<Object>> compile(List<Symbol> arguments) {
+        assert arguments.size() > 1 : "number of arguments must be > 1";
+        String pattern = null;
+        if (arguments.get(1).symbolType() == SymbolType.LITERAL) {
+            Literal<String> literal = (Literal<String>) arguments.get(1);
+            pattern = literal.value();
+            if (pattern == null) {
+                return this;
+            }
+        }
+        String flags = null;
+        if (arguments.size() == 3) {
+            assert arguments.get(2).symbolType() == SymbolType.LITERAL :
+                "3rd argument must be a " + SymbolType.LITERAL;
+            flags = ((Literal<String>) arguments.get(2)).value();
+        }
+        if (pattern != null) {
+            return new MatchesFunction(
+                signature, boundSignature, Pattern.compile(pattern, parseFlags(flags)));
+        }
+        return this;
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input[] args) {
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
+
+        String value = (String) args[0].value();
+        String pattern = (String) args[1].value();
+        if (value == null || pattern == null) {
+            return () -> Collections.emptyIterator();
+        }
+        String flags = null;
+        if (args.length == 3) {
+            flags = (String) args[2].value();
+        }
+
+        if (this.pattern == null) {
+            this.pattern = Pattern.compile(pattern, parseFlags(flags));
+        }
+        Matcher matcher = this.pattern.matcher(value);
+        List<List<String>> rowGroups;
+        if (isGlobal(flags)) {
+            rowGroups = Stream
```

Comment:
is it possible to replace this stream with something like:

```
var group = groups(matcher);
while (group != null) {
    rowGroups.add(...)
    group = groups(matcher)
}
```

Reply:
yes, I was proud of that snippet!